File: alica_engine/src/engine/modelmanagement/factories/RoleFactory.cpp
#include "engine/modelmanagement/factories/RoleFactory.h"
#include "engine/model/Role.h"
#include "engine/modelmanagement/Strings.h"

namespace alica
{
Role* RoleFactory::create(const YAML::Node& roleNode, RoleSet* roleSet)
{
    Role* role = new Role(roleSet);
    Factory::setAttributes(roleNode, role);
    Factory::storeElement(role, alica::Strings::role);

    if (Factory::isValid(roleNode[alica::Strings::taskPriorities])) {
        const YAML::Node& taskPriorities = roleNode[alica::Strings::taskPriorities];
        for (YAML::const_iterator it = taskPriorities.begin(); it != taskPriorities.end(); ++it) {
            Factory::roleTaskReferences.push_back(std::make_tuple(role->getId(), Factory::getReferencedId((*it).first), ((*it).second).as<double>()));
        }
    }

    return role;
}

void RoleFactory::attachReferences()
{
    for (std::tuple<int64_t, int64_t, double> triple : Factory::roleTaskReferences) {
        Role* role = (Role*) Factory::getElement(std::get<0>(triple));
        Task* task = (Task*) Factory::getElement(std::get<1>(triple));
        role->setPriority(task, std::get<2>(triple));
    }
    Factory::planTypePlanReferences.clear();
}
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/RoleFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.Role;
import de.unikassel.vs.alica.planDesigner.alicamodel.Task;
import javafx.util.Pair;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import java.util.HashMap;

public class RoleFactory extends Factory {
    public static void create(Element roleDefinitionSetNode, ConversionProcess cp) {
        NodeList roleNodes = roleDefinitionSetNode.getElementsByTagName(Tags.ROLES);
        for (int i = 0; i < roleNodes.getLength(); i++) {
            Element roleNode = (Element) roleNodes.item(i);
            Role role = new Role();
            Factory.setAttributes(roleNode, role);
            cp.addElement(role);
        }
    }

    public static void attachReferences(ConversionProcess cp) {
        for (HashMap.Entry<Long, Pair<Long, Float>> entry : cp.roleTaskReferences.entrySet()) {
            Role role = (Role) cp.getElement(entry.getKey());
            Task task = (Task) cp.getElement(entry.getValue().getKey());
            if (task == null) {
                throw new RuntimeException("[RoleFactory] Did not find a Taskrepository that contains a task with ID '" + entry.getValue().getKey() + "', but it is referenced by priorities for role with ID '" + role.getId() + "'");
            }
            role.addTaskPriority(task, entry.getValue().getValue());
        }
        cp.roleTaskReferences.clear();
    }
}
